### ISSUE TITLE
HoughCircles sometimes returns [0,0,0]

### ISSUE THREAD

**Issue by knumat** [opencv/opencv#14432]
##### System information (version)
- OpenCV => 4.0.1 (precompiled binary distributed by homebrew)
- Operating System / Platform => macOS Mojave 10.14.4
- Compiler => clang 10.0.0

##### Detailed description
When running HoughCircles on a series of images, it usually works well, but sometimes it will return a single value of [0,0,0] instead of the actual result. I receive an empty vector back on images where no circles are found, so I suspect that a circle was indeed detected, but there was an internal bug.  Unless there is an (undocumented) reason, HoughCircles should return the actual result instead of [0,0,0].

I am calling HoughCircles with a 480x640 CV_8UC1 image (either generated or loaded from a PNG using IMREAD_GRAYSCALE).  Full source code is attached, but the call to HoughCircles is as follows:
```HoughCircles(img, results, HOUGH_GRADIENT, 2, 5, 185, 100, 20, 100);```

The same issue is reported here:
https://stackoverflow.com/questions/53001840/opencv-houghcircles-occasionally-returning-0-0-0

This one seems similar also, except that it returns [10,0,0,0]:
https://stackoverflow.com/questions/49675692/opencv-houghcircles-returns-invalid-circle-parameters

##### Steps to reproduce

1. Download [hough_bug.cpp.txt](https://github.com/opencv/opencv/files/3129029/hough_bug.cpp.txt)
2. Rename the file to hough_bug.cpp
3. Compile hough_bug.cpp against OpenCV 4.0.1 to create a binary named `hough_bug`.
4. Run `./hough_bug` to generate a series of 1000 images, 26 of which demonstrate the issue (for me).  The random seed is fixed so the program will create the same series of images each time you run it.
5. Run `./hough_bug *.png` to test the images generated in the previous step. Again, 26 of the images will demonstrate the issue.  This proves that circle() and GaussianBlur() are not causing the issue.

The console output from an example run is attached:
[hough_bug_output.txt](https://github.com/opencv/opencv/files/3129000/hough_bug_output.txt)

This image demonstrates the bug - returning [0,0,0] instead of a list of valid circles.
![000-bug](https://user-images.githubusercontent.com/29646487/56923279-c89b4f80-6a8f-11e9-8694-1ba2c583d8e4.png)

HoughCircles returns an empty list for this image, since no circles were found (this is expected with these parameters).
![001-normal](https://user-images.githubusercontent.com/29646487/56923280-c89b4f80-6a8f-11e9-998b-a666efb8fec3.png)

HoughCircles returns a list of 2 circles that were detected in this image.
![002-found-two-circles](https://user-images.githubusercontent.com/29646487/56924129-d05bf380-6a91-11e9-8709-4e2cc6652a64.png)

**Comment by alalek** [comment]
Please capture passed input to HoughCircles (image + parameters + code (types of parameters)).
Without reproducer it is almost impossible to fix issue.

**Comment by knumat** [comment]
@alalek It took a bit longer than I expected to create a simple example program to demonstrate this issue.  Could you take a look at the updated description and let me know if you have any idea what could be causing this?  Thanks for the help!

**Comment by alalek** [comment]
@knumat Thank you for the reproducer!
Could you try this patch: #14452?

**Comment by knumat** [comment]
@alalek Thanks for the super-fast turn around on this!  I built OpenCV 4.0.1 from source, and applying this patch resolves the issue for me.  Now HoughCircles returns an empty list for the images that previously returned a list with a single [0,0,0] element.

Quick question...  I noticed that this patch will be applied to the `3.4` branch.  Will this patch also be applied to the `master` branch so that it is included in the next 4.x release?  (I'm not clear about the merge/release strategy on the opencv project.)

Thanks again for the help on this!
